src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectOptionsProcessor.cs

```diff
@@ -137,53 +134,67 @@ private void UpdateProjectOptions_NoLock()
 
             // Override the default documentation mode.
-            var documentationMode = commandLineArgs.DocumentationPath != null ? DocumentationMode.Diagnose : DocumentationMode.Parse;
-            var parseOptions = commandLineArgs.ParseOptions
+            var documentationMode = _commandLineArgumentsForCommandLine.DocumentationPath != null ? DocumentationMode.Diagnose : DocumentationMode.Parse;
+            var parseOptions = _commandLineArgumentsForCommandLine.ParseOptions
                 .WithDocumentationMode(documentationMode);
 
             // We've computed what the base values should be; we now give an opportunity for any host-specific settings to be computed
             // before we apply them
-            compilationOptions = ComputeCompilationOptionsWithHostValues(compilationOptions, _ruleSetFile?.Target.Value);
+            compilationOptions = ComputeCompilationOptionsWithHostValues(compilationOptions, this._ruleSetFile?.Target.Value);
             parseOptions = ComputeParseOptionsWithHostValues(parseOptions);
 
             // For managed projects, AssemblyName has to be non-null, but the command line we get might be a partial command line
             // and not contain the existing value. Only update if we have one.
-            _project.AssemblyName = commandLineArgs.CompilationName ?? _project.AssemblyName;
-
-            var outputAssemblyFilePath = commandLineArgs.GetOutputFilePath(commandLineArgs.OutputFileName);
-            var outputPdbFilePath = commandLineArgs.EmitPdbFile ? commandLineArgs.GetPdbFilePath(commandLineArgs.OutputFileName) : null;
-
+            _project.AssemblyName = _commandLineArgumentsForCommandLine.CompilationName ?? _project.AssemblyName;
             _project.CompilationOptions = compilationOptions;
-            _project.IntermediateOutputFilePath = outputAssemblyFilePath;
             _project.ParseOptions = parseOptions;
 
-            _workspace.SetCompilationOutputs(_project.Id, new CompilationOutputFiles(outputAssemblyFilePath, outputPdbFilePath));
+            var outputFileName = _commandLineArgumentsForCommandLine.OutputFileName;
+            if (outputFileName != null)
+            {
+                // output file path was specified on the command line:
+                var outputFilePath = _commandLineArgumentsForCommandLine.GetOutputFilePath(outputFileName);
+                _project.IntermediateOutputFilePath = outputFilePath;
+                SetCompilationOutputs_NoLock(outputFilePath);
+            }
+            else if (_project.IntermediateOutputFilePath != null)
+            {
+                // output file was not specified on the command line, could have been set previously via SetOutputFileName
+                SetCompilationOutputs_NoLock(_project.IntermediateOutputFilePath);
+            }
+        }
+
+        internal void SetOutputFileName(string outputFilePath)
```

Comment:
It seems odd to me this isn't explicitly holding onto the OutputFileName here that was set. Consider the following series of calls:

1. SetOutputFileName is called; this updates SetCompilationOutputs.
2. UpdateProjectForNewHostValues() is called. This calls through and calls SetCompilationOutputs_NoLock which is once again using the value from _commandLineArgumentsForCommandLine, which may nor may not be correct.

It feels like this really should be more like the ExplicitRuleSetFile property: there's an explicit override that we just take as overriding the parsed command line option.

Reply:
Changed the impl.